akka-stream/src/main/scala/akka/stream/impl/fusing/Fusing.scala

```diff
@@ -0,0 +1,362 @@
+    System.arraycopy(outOwnersB2, 0, outOwners, connStart, conns)
+    copyToArray(outOwnersB3.iterator, outOwners, outStart)
+
+    // FIXME attributes should contain some naming info and async boundary where needed
+    GraphModule(new GraphInterpreter.GraphAssembly(stages, attributes, ins, inOwners, outs, outOwners), shape, Attributes.none, matValIDs)
+  }
+
+  @tailrec private def copyToArray[T](it: ju.Iterator[T], array: Array[T], idx: Int): Int =
+    if (it.hasNext) {
+      array(idx) = it.next()
+      copyToArray(it, array, idx + 1)
+    } else idx
+
+  private def descend(m: Module,
+                      inheritedAttributes: Attributes,
+                      struct: BuildStructuralInfo,
+                      openGroup: ju.Set[Module],
+                      indent: String): MaterializedValueNode = {
+    def log(msg: String): Unit = println(indent + msg)
+    val async = m match {
+      case _: GraphStageModule ⇒ m.attributes.contains(AsyncBoundary)
+      case _ if m.isAtomic     ⇒ true
+      case _                   ⇒ m.attributes.contains(AsyncBoundary)
+    }
+    if (Debug) log(s"entering ${m.getClass} (async=$async, name=${m.attributes.nameLifted})")
+    val localGroup =
+      if (async) struct.newGroup(indent)
+      else openGroup
+
+    if (m.isAtomic) {
+      if (Debug) log(s"atomic module $m")
+      struct.addModule(m, localGroup, inheritedAttributes, indent)
+    } else {
+      val attributes = inheritedAttributes and m.attributes
+      m match {
+        case CopiedModule(shape, _, copyOf) ⇒
+          val ret = descend(copyOf, attributes, struct, localGroup, indent + "  ")
+          struct.rewire(copyOf.shape, shape, indent)
+          ret
+        case _ ⇒
+          struct.enterMatCtx()
+          val subMat: Predef.Map[Module, MaterializedValueNode] =
+            m.subModules.map(sub ⇒ sub -> descend(sub, attributes, struct, localGroup, indent + "  "))(collection.breakOut)
+          val down = m.downstreams.toSet -- m.subModules.map(_.downstreams).reduce(_ ++ _)
+          down.foreach {
+            case (start, end) ⇒ struct.wire(start, end, indent)
+          }
+          val newMat = rewriteMat(subMat, m.materializedValueComputation)
+          val matSrcs = struct.exitMatCtx()
+          matSrcs.foreach { c ⇒
+            if (Debug) log(s"materialized value source: ${struct.hash(c)}")
+            val ms = c.copyOf match {
+              case g: GraphStageModule ⇒ g.stage.asInstanceOf[MaterializedValueSource[Any]]
+            }
+            if (Debug) require(find(ms.computation, m.materializedValueComputation), s"mismatch:\n  ${ms.computation}\n  ${m.materializedValueComputation}")
+            val replacement = CopiedModule(c.shape, c.attributes, new MaterializedValueSource[Any](newMat, ms.out).module)
+            struct.replace(c, replacement, localGroup)
+          }
+          newMat
+      }
+    }
+  }
+
+  private def find(m1: MaterializedValueNode, m2: MaterializedValueNode): Boolean =
```

Comment:
give better names to arguments since the case is not symmetric (m1 and m2 are treated differently, therefore they have a clear role)